Reduce size of code emitted for stack loads/stores from Air on ARM

eugeneia opened the pull request:
#### 6d88e2681672a4fac3e20dca5f30a253ea51f2af
<pre>
Reduce size of code emitted for stack loads/stores from Air on ARM
<a href="https://bugs.webkit.org/show_bug.cgi?id=252469">https://bugs.webkit.org/show_bug.cgi?id=252469</a>

Reviewed by Yusuke Suzuki.

This changes callFrameAddr (AirAllocateRegistersAndStackAndGenerateCode.cpp)
and its users to (which ever succeeds first)

 1. generate an address offset from GPRInfo::callFrameRegister
    (unchanged)
 2. generate an address offset from MacroAssembler::stackPointerRegister
 3. load the offset into extendedOffsetAddrRegister and generate
    a BaseIndex [GPRInfo::callFrameRegister, extendedOffsetAddrRegister]
 4. resort to computing the absolute address in extendedOffsetAddrRegister
    (this was the previous fallback after 1.)

For 3. to work reliably we need to change Air::Arg::isValidForm
to reject BaseIndex addressing when it is not supported on ARM.

The result of these changes is a significant reduction in code size
due to avoiding the fallback behavior in many cases (4).

Code generated for the JetStream2 WASM benchmarks is ~30% smaller on ARMv7
and ~40% smaller on ARM64.

* Source/JavaScriptCore/b3/air/AirAllocateRegistersAndStackAndGenerateCode.cpp:
(JSC::B3::Air::callFrameAddr):
(JSC::B3::Air::GenerateAndAllocateRegisters::flush):
(JSC::B3::Air::GenerateAndAllocateRegisters::alloc):
* Source/JavaScriptCore/b3/air/AirArg.h:
(JSC::B3::Air::Arg::isValidAddrForm):
(JSC::B3::Air::Arg::isValidForm const):

Canonical link: <a href="https://commits.webkit.org/260680@main">https://commits.webkit.org/260680@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/0a55d140bc5d85390f6b79a6f8e8afe3bde9b010

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/109036 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/18115 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/41857 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/87/builds/569 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/118304 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/19573 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/9417 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/101273 "Built successfully") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/114795 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/14672 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/97910 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/42803 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/96644 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/29557 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/84551 "Found 2 new API test failures: /TestWebKit:WebKit.OnDeviceChangeCrash, /TestWebKit:WebKit.PrivateBrowsingPushStateNoHistoryCallback (failure)") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/98076 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/10901 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/30904 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/98868 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/83/builds/8993 "Built successfully and passed tests") | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/11641 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/7834 "Passed tests") | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/46/builds/31018 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/17017 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/50502 "Passed tests") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/106532 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/13243 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/26401 "Passed tests") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/75/builds/4031 "Built successfully and passed tests") | | | | 
<!--EWS-Status-Bubble-End-->

Constellation commented on Source/JavaScriptCore/b3/air/AirAllocateRegistersAndStackAndGenerateCode.cpp:
```diff
@@ -143,19 +143,39 @@ void GenerateAndAllocateRegisters::insertBlocksForFlushAfterTerminalPatchpoints(
+    if (addr.isValidForm(opcode, width))
+        return addr;
+
+    if (isARM64() || isARM_THUMB2()) {
```
We should not have arch condition here. If this is only allowed for these archs, then isValidForm should say false to wrong archs instead.

eugeneia replied:
I removed those in https://github.com/WebKit/WebKit/commit/da11283e6bd1aaa4c73f84c839f1a7b72f508ef3

Constellation commented on Source/JavaScriptCore/b3/air/AirAllocateRegistersAndStackAndGenerateCode.cpp:
```diff
@@ -143,19 +143,39 @@ void GenerateAndAllocateRegisters::insertBlocksForFlushAfterTerminalPatchpoints(
+    
     GPRReg reg = extendedOffsetAddrRegister();
     jit.move(CCallHelpers::TrustedImmPtr(offsetFromFP), reg);
+    if (isARM64() || isARM_THUMB2()) {
```
Ditto

Constellation commented on Source/JavaScriptCore/b3/air/AirAllocateRegistersAndStackAndGenerateCode.cpp:
```diff
@@ -211,14 +239,22 @@ ALWAYS_INLINE void GenerateAndAllocateRegisters::alloc(Tmp tmp, Reg reg, Arg::Ro
             ASSERT(m_map[tmp].spillSlot->byteSize() == bytesForWidth(Width64));
-            m_jit->loadDouble(callFrameAddr(Air::MoveDouble, *m_jit, offset), reg.fpr());
+            auto src = callFrameAddr(Air::MoveDouble, *m_jit, offset, m_code.frameSize(), Width64);
+            ASSERT(src.isAddr());
```
Can you explain why this is correct?

eugeneia replied:
It used to be an invariant because this patch `isValidForm` rejected BaseIndex addresses for Air::MoveDouble. I've changed `isValidIndexForm` in https://github.com/WebKit/WebKit/commit/da11283e6bd1aaa4c73f84c839f1a7b72f508ef3 to only reject  Air::MoveDouble and Air::MoveFloat on ARM_THUMB2 (because the MacroAssemblerARMv7 will try to use the addressTempRegister to materialize them.) Hence, I've replaced the ASSERTs with branches to handle BaseIndex addresses for loadDouble/storeDouble/loadVector/storeVector (those can and do occur on ARM64.)

Constellation commented on Source/JavaScriptCore/b3/air/AirAllocateRegistersAndStackAndGenerateCode.cpp:
```diff
@@ -211,14 +239,22 @@ ALWAYS_INLINE void GenerateAndAllocateRegisters::alloc(Tmp tmp, Reg reg, Arg::Ro
             ASSERT(m_map[tmp].spillSlot->byteSize() == bytesForWidth(Width128));
-            m_jit->loadVector(callFrameAddr(Air::MoveDouble, *m_jit, offset), reg.fpr());
+            auto src = callFrameAddr(Air::MoveDouble, *m_jit, offset, m_code.frameSize(), Width128);
+            ASSERT(src.isAddr());
```
Ditto

Constellation commented on Source/JavaScriptCore/b3/air/AirArg.h:
```diff
@@ -1396,7 +1412,7 @@ class Arg {
         case CallArg:
             return isValidAddrForm(opcode, offset(), width);
         case Index:
-            return isValidIndexForm(scale(), offset(), width);
```
isValidIndexForm should also take op ode instead

eugeneia replied:
Fixed in https://github.com/WebKit/WebKit/commit/da11283e6bd1aaa4c73f84c839f1a7b72f508ef3

Constellation commented on Source/JavaScriptCore/b3/air/AirAllocateRegistersAndStackAndGenerateCode.cpp:
```diff
@@ -211,14 +239,26 @@ ALWAYS_INLINE void GenerateAndAllocateRegisters::alloc(Tmp tmp, Reg reg, Arg::Ro
         } else {
             ASSERT(m_map[tmp].spillSlot->byteSize() == bytesForWidth(Width128));
-            m_jit->loadVector(callFrameAddr(Air::MoveDouble, *m_jit, offset), reg.fpr());
+            auto src = callFrameAddr(Air::MoveDouble, *m_jit, offset, m_code.frameSize(), Width128);
```
It looks like this is pre-existing issue, but we should use `MoveVector` (but it does not matter in terms of correctness I think).

Constellation commented on Source/JavaScriptCore/b3/air/AirAllocateRegistersAndStackAndGenerateCode.cpp:
```diff
@@ -175,14 +191,26 @@ ALWAYS_INLINE void GenerateAndAllocateRegisters::flush(Tmp tmp, Reg reg)
     } else {
         ASSERT(m_map[tmp].spillSlot->byteSize() == bytesForWidth(Width128));
-        m_jit->storeVector(reg.fpr(), callFrameAddr(Air::MoveDouble, *m_jit, offset));
+        auto dest = callFrameAddr(Air::MoveDouble, *m_jit, offset, m_code.frameSize(), Width128);
```
Ditto.

Constellation commented on Source/JavaScriptCore/b3/air/AirAllocateRegistersAndStackAndGenerateCode.cpp:
```diff
@@ -143,19 +143,35 @@ void GenerateAndAllocateRegisters::insertBlocksForFlushAfterTerminalPatchpoints(
+        return addr;
+
+    // Try finding a valid offset from the stack pointer register instead.
+    // (this trick is also done in Air::lowerStackArgs:126)
```
Let's remove this line since this line number etc. is fragile.

Constellation reviewed (approved):
r=me

Constellation commented on Source/JavaScriptCore/b3/B3LowerToAir.cpp:
```diff
@@ -571,7 +571,7 @@ class LowerToAir {
 
             auto tryIndex = [&] (Value* index, Value* base) -> Arg {
-                std::optional<unsigned> scale = scaleForShl(index, offset, width);
+                std::optional<unsigned> scale = scaleForShl(Air::Move, index, offset, width);
```
Is `Move` correct?

eugeneia replied:
I think so? I chose `Air::Move` because earlier in the function there is a comment:
```
// This function currently is currently only used for loads/stores, so
// using Air::Move is appropriate.
```
(and a similar choice).

Not sure though. I find it tricky to do it these mappings from Air opcodes to desired `isValidForm` behavior.

Constellation commented on Source/JavaScriptCore/b3/B3LowerToAir.cpp:
```diff
@@ -598,7 +598,7 @@ class LowerToAir {
             // We avoid using baseless indexes because their encoding isn't particularly efficient.
             if (m_locked.contains(left) || !address->child(1)->isInt32(1)
-                || !Arg::isValidIndexForm(1, offset, width))
+                || !Arg::isValidIndexForm(Air::Move, 1, offset, width))
```
Is `Move` correct?

eugeneia replied:
See above.

Constellation commented on Source/JavaScriptCore/b3/B3LowerToAir.cpp:
```diff
@@ -613,7 +613,7 @@ class LowerToAir {
             WasmAddressValue* wasmAddress = address->as<WasmAddressValue>();
             Value* pointer = wasmAddress->child(0);
-            if (!Arg::isValidIndexForm(1, offset, width) || m_locked.contains(pointer))
+            if (!Arg::isValidIndexForm(Air::Move, 1, offset, width) || m_locked.contains(pointer))
```
Is `Move` correct?

eugeneia replied:
See above.

Constellation commented on Source/JavaScriptCore/b3/B3LowerToAir.cpp:
```diff
@@ -2325,7 +2325,7 @@ class LowerToAir {
         
         auto tryShl = [&] (Value* shl, Value* other) -> bool {
-            std::optional<unsigned> scale = scaleForShl(shl, offset);
+            std::optional<unsigned> scale = scaleForShl(leaOpcode, shl, offset);
```
Is `leaOpcode` correct?

eugeneia replied:
Again, I think so? Not sure what the other candidates would be?

eugeneia committed:
Reduce size of code emitted for stack loads/stores from Air on ARM
https://bugs.webkit.org/show_bug.cgi?id=252469

Reviewed by Yusuke Suzuki.

This changes callFrameAddr (AirAllocateRegistersAndStackAndGenerateCode.cpp)
and its users to (which ever succeeds first)

 1. generate an address offset from GPRInfo::callFrameRegister
    (unchanged)
 2. generate an address offset from MacroAssembler::stackPointerRegister
 3. load the offset into extendedOffsetAddrRegister and generate
    a BaseIndex [GPRInfo::callFrameRegister, extendedOffsetAddrRegister]
 4. resort to computing the absolute address in extendedOffsetAddrRegister
    (this was the previous fallback after 1.)

For 3. to work reliably we need to change Air::Arg::isValidForm
to reject BaseIndex addressing when it is not supported on ARM.

The result of these changes is a significant reduction in code size
due to avoiding the fallback behavior in many cases (4).

Code generated for the JetStream2 WASM benchmarks is ~30% smaller on ARMv7
and ~40% smaller on ARM64.

* Source/JavaScriptCore/b3/air/AirAllocateRegistersAndStackAndGenerateCode.cpp:
(JSC::B3::Air::callFrameAddr):
(JSC::B3::Air::GenerateAndAllocateRegisters::flush):
(JSC::B3::Air::GenerateAndAllocateRegisters::alloc):
* Source/JavaScriptCore/b3/air/AirArg.h:
(JSC::B3::Air::Arg::isValidAddrForm):
(JSC::B3::Air::Arg::isValidForm const):

Canonical link: https://commits.webkit.org/260680@main
